tests/test_api.py

```diff
@@ -0,0 +1,386 @@
+    "users": {
+        "json": user,
+        "model": Users,
+        "allowed_methods": {"POST", "PATCH", "GET"},
+        "PATCH_json": {"role": CIDCRole.CIMAC_USER.value},
+        "filters": {
+            "empty": {
+                "where": f"role=='{CIDCRole.CIMAC_USER.value}' and email=='{user['email']}'"
+            },
+            "one": {
+                "where": f"role=='{CIDCRole.CIMAC_USER.value}' or email=='{user['email']}'"
+            },
+        },
+        "additional_records": [
+            {**user, "id": 2, "email": "foo@bar.com"},
+            {**user, "id": 3, "email": "fizz@buzz.com"},
+        ],
+    },
+    "trial_metadata": {
+        "json": trial_metadata,
+        "model": TrialMetadata,
+        "allowed_methods": {"POST", "PATCH", "GET"},
+        "lookup_field": "trial_id",
+        "PATCH_json": {"metadata_json": {"foo": "bar"}},
+    },
+    "downloadable_files": {
+        "json": downloadable_file,
+        "model": DownloadableFiles,
+        "allowed_methods": {"GET"},
+        "POST_setup": ["trial_metadata"],
+        "PATCH_json": {"upload_type": "fizzbuzz"},
+        "filters": {
+            "empty": {
+                "where": f"trial_id=='{trial_metadata['trial_id']}' and upload_type=='wes'"
+            },
+            "one": {
+                "where": f"trial_id=='{trial_metadata['trial_id']}' and upload_type=='olink' and id==1"
+            },
+        },
+        "additional_records": [
+            {**downloadable_file, "id": 2, "object_url": "foo/bar"},
+            {**downloadable_file, "id": 3, "object_url": "fizz/buzz"},
+        ],
+    },
+    "permissions": {
+        "json": permission,
+        "model": Permissions,
+        "allowed_methods": {"POST", "PATCH", "GET", "DELETE"},
+        "POST_setup": ["users", "trial_metadata"],
+        "PATCH_json": {"upload_type": "fizzbuzz"},
+        "filters": {
+            "empty": {"where": "granted_to_user==2"},
+            "one": {"where": "granted_to_user==1"},
+        },
+    },
+    "upload_jobs": {
+        "json": upload_job,
+        "model": UploadJobs,
+        "allowed_methods": {"PATCH", "GET"},
+        "POST_setup": ["users", "trial_metadata"],
+        "PATCH_json": {"upload_type": "fizzbuzz"},
+        "mocks": [
+            lambda monkeypatch: monkeypatch.setattr(
+                "cidc_api.services.ingestion.gcloud_client.revoke_upload_access",
+                MagicMock(),
+            )
+        ],
+    },
+}
+
+
+def assert_dict_contains(base, target):
+    assert isinstance(target, dict) and isinstance(base, dict)
+    for key, value in target.items():
+        assert key in base
+        assert base[key] == value or isinstance(value, datetime)
+
+
+def setup_mocks(config, monkeypatch):
+    if "mocks" in config:
+        for mock in config["mocks"]:
+            mock(monkeypatch)
+
+
+@pytest.mark.parametrize("resource, config", resource_requests.items())
```

Comment:
i'd add `ids` fn that will give human readable descriptions to the parameter set:
```suggestion
@pytest.mark.parametrize("resource, config", resource_requests.items(), ids=lambda x: x if isinstance(x, str) else x['model'])
```

Reply:
The auto-generated parameter labels seem friendly enough, IMO:
```
tests/test_api.py::test_resource_post[new_users-config0]
tests/test_api.py::test_resource_post[users-config1]
tests/test_api.py::test_resource_post[trial_metadata-config2]
tests/test_api.py::test_resource_post[downloadable_files-config3]
tests/test_api.py::test_resource_post[permissions-config4]
...
```

Reply:
right, that's friendly !